chroma-manager/chroma_core/plugins/block_devices.py

```diff
@@ -0,0 +1,559 @@
+                etc.
+        '''
+
+        return [
+            value for value in self.table.values()
+            if value.startswith(device_fullpath)
+            ]
+
+    def normalized_device_path(self, device_path):
+        normalized_path = os.path.realpath(device_path)
+
+        # This checks we have a completely normalized path, perhaps the
+        # stack means our current normal path can actually be
+        # normalized further.
+        # So if the root to normalization takes multiple
+        # steps this will deal with it
+        # So if /dev/sdx normalizes to /dev/mmapper/special-device
+        # and /dev/mmapper/special-device normalizes to /dev/md/mdraid1,
+        # then /dev/sdx will normalize to /dev/md/mdraid1
+        # /dev/sdx -> /dev/mmapper/special-device -> dev/md/mdraid1
+
+        # As an additional measure to detect circular references
+        # such as A->B->C->A in
+        # this case we don't know which is the
+        # normalized value so just drop out once
+        # it repeats.
+        visited = set()
+
+        while (normalized_path not in visited) and (
+                    normalized_path in self.table):
+            visited.add(normalized_path)
+            normalized_path = self.table[normalized_path]
+
+        return normalized_path
+
+
+def parse_sys_block(device_map):
+    xs = create_device_list(device_map)
+
+    mutate_parent_prop(xs)
+
+    node_block_devices = reduce(
+        lambda d, x: dict(d, **{x['path']: x['major_minor']}), xs, {})
+
+    block_device_nodes = reduce(
+        lambda d, x: dict(d, **{x['major_minor']: x}), xs, {})
+
+    ndt = NormalizedDeviceTable(xs)
+
+    (ndt, vgs, lvs) = parse_dm_devs(filter(lambda x: x.get('dm_uuid') is not None, xs),
+                                    block_device_nodes,
+                                    ndt)
+
+    (ndt, mds) = parse_mdraid_devs(filter(lambda x: x.get('md_uuid') is not None, xs),
+                                   node_block_devices,
+                                   ndt)
+
+    local_fs = parse_localfs_devs(filter(local_fs_filter, xs),
+                                  node_block_devices,
+                                  ndt)
+
+    return [vgs, lvs, mds, local_fs, block_device_nodes]
+
+
+def parse_zpools(zpool_map, zfs_objs, block_device_nodes):
+    """
+    :param zpool_map: dictionary of pools keyed on guid
+    :param zfs_objs: list of dataset dictionaries
+    :param block_device_nodes: dictionary of block device dictionaries keyed on major minor
+    :return: list of pools datasets and block_device_nodes
+    """
+    zpools = {}
+    datasets = {}
+    # fixme: get zvols from zfs_objs using some identifier to discern from datasets
+
+    for pool in zpool_map.values():
+        name = pool['name']
+        guid = pool['guid']
+
+        size = pool['size']
+        drive_mms = get_drives([child['Disk'] for child in pool['vdev']['Root']['children']],
```

Comment:
This won't be sufficient for mapping all the drives. 

The `VDev` tree is recursive and can contain things like mirrors.